### PR TITLE
Feature spinop

### PR DESCRIPTION
**Pull request by Hadrien-Montanelli**
This branch implements a time-depedent PDE solver in 1D, 2D and 3D. It is based on Fourier spectral methods and exponential integrators.

### CODE REVIEW

**Comment by trefethen** [comment]
If this gets merged by Friday, I'll include a problem using `spin` in my Scientific Computing for DPhil Students handed out that day.  (Even if it doesn't get merged I might still do that with a pointer to the branch, but it's cooler if the code is in development.)

**Comment by aaustin141** [comment]
This comment is a reminder to myself to reinstate the test for the `spin` chapter of the Guide once this has been merged.

**Comment by Hadrien-Montanelli** [comment]
@trefethen, should we wait for `chebfun3`? `spin3` uses `chebfun3` objects.

**Comment by trefethen** [comment]
My current view is that spherefun and spin are nearly ready to go and Chebfun3 isn't, so ideally we shouldn't let the latter hold up the former.  So my inclination would be not to wait for Chebfun3.  Instead, one approach might be for spin3 to issue an error message for the moment that says that it depends on Chebfun3 and that Chebfun3 will be released in a couple of months.

**Comment by jrus** [comment]
Is the code for Chebfun3 or any other information about it anywhere public? (I’m curious to look what if anything you use as a 3d analog of the SVD, and whether it generalizes to 4 dimensions.)

**Comment by Hadrien-Montanelli** [comment]
Ask @bhashemi for the Chebfun3 code.

**Comment by bhashemi** [comment]
@jrus , the code for `Chebfun3` is not yet public, because it is still subject to change. It will be available in the coming months. On the other hand, as you expected, the main idea is a 3D analog of SVD, not in the CP format, but in the Tucker format. Here is a link to how CP looks:
http://www.eecs.qmul.ac.uk/~ioannisp/images/parafac.jpg
and here is a link to how Tucker looks:
http://image.slidesharecdn.com/snlp2013-komachi-130909085508-/95/a-tensorbased-factorization-model-of-semantic-compositionality-10-638.jpg?cb=1378716993

**Comment by Hadrien-Montanelli** [comment]
@jaurentz @mhsnjvd -- I've added the type of the properties in the classdef files.

File: @spinoperator/spinoperator.m [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    %% CLASS PROPERTIES:
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    properties ( Access = public )
+        domain              % Spatial domain (1x2 DOUBLE in 1D, 1x4 in 2D, 
+                            % 1x6 in 3D)
+        init                % Initial condition (CHEBFUN in 1D, CHEBFUN2 in 2D,
+                            % CHEBFUN3 in 3D, CHEBMATRIX for systems)
+        linearPart          % Linear part of the operator (FUNCTION HANDLE)
+        nonlinearPart       % Nonlinear part of the operator (FUNCTION HANDLE)
+        tspan               % Time interval (DOUBLE)
```
Comment (jaurentz) [inline @spinoperator/spinoperator.m]: % Vector of monotonically increasing time samples (DOUBLE)

File: @spinoperator/spinoperator.m [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    end
+    
+    % DEPENDENT PROPERTIES:
+    properties ( Access = public, Dependent = true )
+        numVars             % Number of unknown functions (1x1 DOUBLE)
```
Comment (jaurentz) [inline @spinoperator/spinoperator.m]: (1 x 1 INT)?

File: @spinpreference/spinpreference.m [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    properties ( Access = public )
+        dataToPlot = 'real'; % Which data to plot when complex values (STRING)
+        dealias = 'off';     % To use dealiasing with 2/3-rule (STRING)
+        dt                   % Time-step (1x1 DOUBLE)
+        dtmin = 1e-10;       % Min. time-step for apative time-grid (1x1 DOUBLE)
+        dtmax                % Max. time-step for apative time-grid (1x1 DOUBLE)
+        errTol               % Desired accuracy on the solution (1x1 DOUBLE)
+        iterPlot             % Plot every ITERPLOT iterations (1x1 DOUBLE)
+        M                    % Number of points for complex means (1x1 DOUBLE)
```
Comment (jaurentz) [inline @spinpreference/spinpreference.m]: (1 x 1 INT)?

File: @spinscheme/spinscheme.m [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+% Copyright 2016 by The University of Oxford and The Chebfun Developers.
+% See http://www.chebfun.org/ for Chebfun information.
+    
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    %% CLASS PROPERTIES:
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    properties ( Access = public )
+        
+        stages          % number of internal stages (1x1 DOUBLE)
```
Comment (jaurentz) [inline @spinscheme/spinscheme.m]: (1x1 INT)?

File: @spinoperator/spinoperator.m (hunk 2 above)
Comment (Hadrien-Montanelli) [inline @spinoperator/spinoperator.m]: Mathematically yes, in Matlab it's a double though.

File: @spinpreference/spinpreference.m (hunk 3 above)
Comment (Hadrien-Montanelli) [inline @spinpreference/spinpreference.m]: Same comment.

File: @spinscheme/spinscheme.m (hunk 4 above)
Comment (Hadrien-Montanelli) [inline @spinscheme/spinscheme.m]: Same comment.

**Comment by jaurentz** [comment]
@Hadrien-Montanelli I highlighted some of the class properties that could have a bit more accurate descriptions. Most notably the singleton variables that, although stored as `numeric` in matlab, are really assumed to be positive integers. I'm not sure what the correct thing to do is but I think we should try to be as clear as possible.

File: @spinoperator/spinoperator.m (hunk 1 above)
Comment (Hadrien-Montanelli) [inline @spinoperator/spinoperator.m]: `tspan` can be 1xN where N is the time slices the user specified -- what do you think we should say?

**Comment by Hadrien-Montanelli** [comment]
@jaurentz -- your call. If you think it's clearer to say INT, please go ahead and change it!

File: @spinscheme/spinscheme.m [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+%
+%   MODIFIED GENERALIZED LAWSON: 'modgenlawson41', modgenlawson42', 
+%                                'modgenlawson43', modgenlawson44',
+%                                'modgenlawson45'
+%
+%   PREDICTOR-CORRECTOR: 'pec423', 'pecec433', 'pec524', 'pecec534', 'pec625',
+%                        'pecec635', 'pec726', 'pecec736'
+%
+% See also SPINSCHEME/COMPUTECOEFFS, SPIN, SPIN2, SPIN3.
+
```
Comment (jaurentz) [inline @spinscheme/spinscheme.m]: Reference your review paper at some point.

Comment (Hadrien-Montanelli) [inline @spinscheme/spinscheme.m]: Ok!

File: @spinscheme/startMultistep.m [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    
+    [cNew, NcNew] = oneStep(K, schemeCoeffs, Nc, Nv, nVars, uSol, NuSol);
+     
+    % If adpative in time, two steps in time with DT/2 and N points:
+    if ( adaptiveTime == 1 )
+        [cNew2, NcNew2] = oneStep(K, schemeCoeffs2, Nc, Nv, nVars, uSol, NuSol);
+        [cNew2, NcNew2] = oneStep(K, schemeCoeffs2, Nc, Nv, nVars, cNew2, ...
+            NcNew2);
+        err = max(abs(cNew{1}(:) - cNew2{1}(:)));
+        err = err/max(abs(cNew2{1}(:)));
```
Comment (jaurentz) [inline @spinscheme/startMultistep.m]: Is zero division possible here?

Comment (Hadrien-Montanelli) [inline @spinscheme/startMultistep.m]: Yes, but not sure it would happen in practice.

File: @spinscheme/oneStep.m [hunk 7]
```diff
@@ -0,0 +1,133 @@
+function [uSol, NuSol] = oneStep(K, schemeCoeffs, Nc, Nv, nVars, uSol, NuSol)
+%ONESTEP   Compute solution at t_{n+1} from solution at previous time-steps.
+%   [USOL, NUSOL] = ONESTEP(K, SCHEMECOEFFS, NC, S, USOL, NUSOL) updates the
```
Comment (jaurentz) [inline @spinscheme/oneStep.m]: Comment doesn't match function declaration.

File: @spinscheme/oneStep.m [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+else
+    uSol = {sol, uSol{1:end-1}}.';
+end
+ 
+% Nonlinear evaluation of the solution at t_{n+1}:
+switch K.scheme
+    case {'pec423', 'pecec433', 'pec524', 'pecec534', 'pec625', 'pecec635', ...
+            'pec726', 'pecec736'}
+        Nsol = NvSol{end};
+    otherwise
```
Comment (jaurentz) [inline @spinscheme/oneStep.m]: Does this protect against scheme names that aren't supported? Could a naive developer get bad behavior here using an invalid scheme name?

File: @spinscheme/oneStep.m (hunk 7 above)
Comment (Hadrien-Montanelli) [inline @spinscheme/oneStep.m]: Ok, will change this.

File: @spinscheme/psiEval.m [hunk 9]
```diff
@@ -0,0 +1,21 @@
+function psi = psiEval(l, C, LR, N, dim, nVars)
+%PSIEVAL   Evaluate a psi function.
+%   PSI = PSIEVAL(L, LR, N, DIM, NVARS) evaluates the psi-function of index L
```
Comment (jaurentz) [inline @spinscheme/psiEval.m]: Comments don't match function declaration.

File: @spinscheme/oneStep.m (hunk 8 above)
Comment (Hadrien-Montanelli) [inline @spinscheme/oneStep.m]: Forget about this, this will be removed soon, just needed this for some testing.

File: @spinscheme/psiEval.m (hunk 9 above)
Comment (Hadrien-Montanelli) [inline @spinscheme/psiEval.m]: Ok, thanks

**Comment by jaurentz** [comment]
I believe the reason for having the classes `spinpref`, `spinpref2` and `spinpref3` is so that there can exist separate defaults for 1, 2 and 3 dimensions. I am perfectly fine with this, but I think if this is the goal there should be no defaults in the `spinpreference` class. My proposal is to remove all the defaults from `spinpreference` and have default values only in the `spinpref`, `spinpref2` and `spinpref3` files.

File: @spinoperator/spinoperator.m [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            % Convert FUNCN to a string STRN:
+            strN = func2str(N);
+            
+            % Get the number of variables NVARS:
+            nVars = nargin(N);
+            
+            % Get the dimension DIM:
+            dim = getDimension(S);
+            
+            % For scalar equations in 1D, we support nonlinearities of the form
```
Comment (jaurentz) [inline @spinoperator/spinoperator.m]: How is this enforced at the user level?

File: @spinoperator/spinoperator.m [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+                    diffOrder = diffOrderTwoOrGreater{1}(2:end);
+                elseif ( isempty(diffOrderOne) == 0 )
+                    diffOrder = '1';
+                else
+                    diffOrder = '0';
+                end
+                Nc = ['@(u) diff(u,', diffOrder, ')'];
+                Nc = eval(Nc);
+                
+            % For scalar equations in 2D and 3D, and for systems of equations in
```
Comment (jaurentz) [inline @spinoperator/spinoperator.m]: How is this enforced at the user level?

File: @spinoperator/solvepde.m [hunk 12]
```diff
@@ -0,0 +1,663 @@
...
+end
+
+% Time-step dt:
+dtmin = pref.dtmin;
+dtmax = pref.dtmax;
+adaptiveTime = isempty(pref.dt);
+if ( adaptiveTime == 1 )
+    % Adaptive in time, start with DTMAX: 
+    % (Unless the interval is shorter than DTMAX.)
+    dt = min(tspan(end), dtmax);
```
Comment (jaurentz) [inline @spinoperator/solvepde.m]: Should this be `dt = min(tspan(end)-tspan(1), dtmax)`?

File: @spinoperator/solvepde.m [hunk 13]
```diff
@@ -0,0 +1,663 @@
...
+else
+    % Not adpative in time, i.e., use the dt given by the user:
+    dt = pref.dt;
```
Comment (jaurentz) [inline @spinoperator/solvepde.m]: What happens if `dt` is bigger than the length of `tspan`?

File: @spinoperator/solvepde.m [hunk 14]
```diff
@@ -0,0 +1,663 @@
...
+                    end
+                    v = [v; temp];
+                end
+                valuesUpdated = 1;
+                vWater{iter/iterplot + 1} = v;
+                twater = [twater, t];
+            end
+                      
+            % Output the solution if T correponds to an entry of TSPAN:
+            if ( abs(t - tspan(pos)) < 1e-10 )
```
Comment (jaurentz) [inline @spinoperator/solvepde.m]: Should this be a relative check?

**Pull request merged**